### DISCUSSION TITLE
username display  of api access

### DISCUSSION THREAD

**Discussion by rockonedege** [requarks/wiki#6140] [Error / Bug Report]
I create a group called `bots`  and a user  `bot` inside it, then I log in as `user` and created an API access token, which is used to update a page through  a python script.

I am expecting to see that the page is displayed as being updated by `bot`, instead it shows the user I used to create `bot`.

then I manually edit the page, and it shows being updated by `bot` as expected.

How can I achieve it through a script?

**Comment by NGPixel** [comment]
Creating a bot user is pointless as API tokens are not tied to a user and will always act as the Administrator user, but with the permissions you gave the token.

**Comment by rockonedege** [comment]
This behavior kinda prevented us from using wiki this way: using a script to convert Excel Spreadsheets into markdown text and pubish/update onto wiki. We'd like to have the real author names to be logged.

Do you think it's a feature worthy of requesting? I could submit one if so.

**Comment by mittkkoo** [comment]
> I create a group called `bots` and a user `bot` inside it, then I log in as `user` and created an API access token, which is used to update a page through a python script.
> 
> I am expecting to see that the page is displayed as being updated by `bot`, instead it shows the user I used to create `bot`.
> 
> then I manually edit the page, and it shows being updated by `bot` as expected.
> 
> How can I achieve it through a script?

This is a highly interesting workflow! Would you like to share here the python script, which converts an Excel file in markdown and updates a wiki.js pages?

**Reply by rockonedege** [reply]
Sure, 
## converting excel to markdown
there are many ways, after I ended up not using Python but two external a tools 
- https://github.com/inosion/madato to convert Excel to Json
- https://pypi.org/project/chevron/ to apply the above json to a mustache template

an example of graphql mustache
```mustache
{
    pages {
        single(id: {{{_id}}}) {
            id
            path
            title
            locale
            description
            createdAt
            creatorName
            creatorEmail
            updatedAt
            authorName
            authorEmail
            contentType
            content
            tags {
                tag
                title      
              }
        }
    }
}
```

## wikijs API with Python
first, use https://insomnia.rest/ to test you graphql api. The best feature is it can generate code snippet in a language of your choice once it works.  Below is an improved version based on its generated code:
```python
import json
import http.client
from dotenv import dotenv_values


def query(payload: str, server: str = "wiki.advme.global"):
    """
    run one graphQL query against wiki.advme.global
    - load authentication info from .env file 
    """
    conn = http.client.HTTPConnection(server)

    payload = json.dumps(dict(query=payload))

    # take environment variables from .env.
    authorization = dotenv_values(".env")['authorization']

    headers = {
        'Content-Type': "application/json",
        'Authorization': authorization
    }

    conn.request("POST", "/graphql", payload, headers)

    res = conn.getresponse()
    data = res.read().decode("utf-8")
    data = json.loads(data)

    # pprint(data)

    return data

```
the `.env` file is something like this
```
authorization=Bearer balabala......
```
